compiler/Makefile

```diff
@@ -1,18 +1,22 @@
 COMPILER_SRC=$(wildcard *.bal) $(wildcard modules/*/*.bal)
-COMPILER_JAR=target/bin/nballerina.jar
+COMPILER_OUT=../build
+COMPILER_JAR=$(COMPILER_OUT)/bin/nballerina.jar
 BAL?=bal
 BAL_BUILD_FLAGS ?= --offline
 
 all: $(COMPILER_JAR)
 
-test:
-	$(BAL) build $(BAL_BUILD_FLAGS)
+# `bal test` dosn't produce a jar, but `test` target of `test/Makefile` needs it
+# therefor this depends on `$(COMPILER_JAR)`
+test: $(COMPILER_JAR)
```

Comment:
Better to handle this in the top-level Makefile: the `test` target of this make does not depend on the compiler jar.

Here's how I think it should work to minimize total elapsed time with lots of cores:

Top-level `make all` should do `make all` in compiler and runtime in parallel.

Top-level `make test` should do sequentially
1. `make all`
2. `make test` in runtime, compiler and test in parallel

I also think it's better for `make test` to use the different directory (something like `compiler/testTarget`) -- that way both `bal build` and `bal test` will produce something that `../test` can use if we want to run things manually.

Reply:
Ok, let me change it that way. But if beta5 starts wiring the jar during `bal test`, we'll have to change back.
`bal test --target-dir` don't seem to work, I'll create an issue.
It might be good to change the default target dir via `[build-options]` in `Ballerina.toml`, I'll experiment and see.

Reply:
If `bal test --target-dir` doesn't work, then we will have to use `bal build --target-dir`.  Only problem is that means we usually will have to use `make` rather than calling `bal build`.  `make` in /test should use the jar produced by `make all` in /compiler, so that `make test` in /compiler can run in parallel with `make` in /test

Reply:
I deviated slightly form the above suggestion. Top-level `make test` will run `make all` only on compiler and runtime. Otherwise all tests get compiled first. It didn't feel right to spending time compiling higher subsets when there is an error in the lower.

I couldn't used `[build-options]` since that will be used by the `bal test` and clean the dir.

Reply:
Current plan form jBallerina team (@azinneera) is to fix the jar deletion issue in beta5 but not the ` --target-dir` missing in `bal test` issue [(#33954)](https://github.com/ballerina-platform/ballerina-lang/issues/33954). Since `[build-options]` is already working for all, including `bal test`, I believe this is good enough for us.

We'll not be able to have different build dirs for `test` and `build`, but as long as `test` doesn't override the jar, it shouldn't be an issue.

Reply:
Does `[build-options]` have to be specified in a file in the source directory?  If so, that's not really enough, since we want to be able to build in a user-specified build directory without changing anything in the source directory.

Reply:
Yes, it need to be specified in `Ballerina.toml`. I imagined being able to specify a fixed dir outside source is enough for us to implement #664. Are you planing something beyond that?

Reply:
Build systems (e.g. cmake, autotools) usually allow the user to build into a directory they choose at build time, without touching the source directory.  This means in particular you can have multiple build trees from the same source tree. So, no, it's not enough for #664.  When we do #664, we should do it once, properly.